### ISSUE TITLE
Handling overloads on generics with different arguments versus Unknown

### ISSUE THREAD

**Issue by Dr-Irv** [microsoft/pyright#5216]
Referencing https://github.com/microsoft/pyright/issues/5178#issuecomment-1566190533 where @erictraut suggested that it might be possible to remove `TimestampSeries`, `TimedeltaSeries`, etc. in `pandas-stubs` by using declarations such as `def __add__(self: Series[Timestamp], ...)`

Here is an example that illustrates why I created the specialized types.

```python
from __future__ import annotations
from typing import Generic, TypeVar, Self, overload, reveal_type
import datetime as dt

T = TypeVar("T", bound=dt.datetime | dt.timedelta)


class Holder(Generic[T]):
    @overload
    def __new__(cls, data: list[T]) -> Self:
        ...

    @overload
    def __new__(cls, data: list) -> Self:
        ...

    def __new__(cls, data: list) -> Self:
        result = object.__new__(cls)
        result._data = [i for i in data]  # type: ignore
        return result

    def __repr__(self) -> str:
        return str(self._data)  # type: ignore

    @overload
    def __sub__(
        self: Holder[dt.datetime], other: dt.timedelta | Holder[dt.timedelta]
    ) -> Holder[dt.datetime]:
        ...

    @overload
    def __sub__(
        self: Holder[dt.datetime], other: dt.datetime | Holder[dt.datetime]
    ) -> Holder[dt.timedelta]:
        ...

    @overload
    def __sub__(
        self: Holder[dt.timedelta], other: dt.timedelta | Holder[dt.timedelta]
    ) -> Holder[dt.timedelta]:
        ...

    @overload
    def __sub__(self, other: float | list[float]) -> Holder:
        ...

    def __sub__(
        self,
        other: dt.timedelta
        | Holder[dt.timedelta]
        | dt.datetime
        | Holder[dt.datetime]
        | float
        | list[float],
    ) -> Holder:
        if isinstance(other, dt.timedelta | dt.datetime | float):
            ndata = [i - other for i in self._data]  # type: ignore
        elif isinstance(other, Holder):
            ndata = [i - j for i, j in zip(self._data, other._data)]  # type: ignore
        else:
            raise ValueError("invalid arg to __sub__")
        return Holder(ndata)


h1 = Holder([dt.datetime(2023, 3, d) for d in [15, 16, 17]])
h2 = Holder([dt.datetime(2023, 2, d) for d in [15, 18, 20]])


print(h1)
print(h2)

s1 = h1 - h2
reveal_type(s1)
print(s1)

hi1 = Holder([10, 11, 12])
hi2 = Holder([1, 4, 8])

reveal_type(hi1)
reveal_type(hi2)

si1 = hi1 - hi2
reveal_type(si1)  # pyright reports Holder[datetime], but should be Holder[Unknown]
print(si1)


```

We need to report `Series[Unknown]` when the type of `Holder` (mimicing `Series` in this case) is `Holder[Unknown]`

If I created a separate `TimestampHolder` and `TimedeltaHolder` then things work fine.

Also, need it to work with `mypy`, although `mypy` is requiring the type declarations on `h1`, `h2`, `hi1`, `hi2`, if you make the second overload for `__new__()`  return `Self`, whereas that is _not_ required if the second overload for `__new__()` returns `Holder` .

**Comment by erictraut** [comment]
It looks like mypy handles things correctly if you switch from `__new__` to `__init__`.

```python
class Holder(Generic[T]):
    @overload
    def __init__(self, data: list[T]) -> None:
        ...

    @overload
    def __init__(self: Holder, data: list) -> None:
        ...

    def __init__(self, data: list) -> None:
        super().__init__()
        self._data = [i for i in data]
```

I'll consider the idea of evaluating `Holder[Unknown]` consistent with mypy when the overload is ambiguous due to an `Unknown` or `Any` type argument. I want to research mypy's exact algorithm here to make sure that it's an intended behavior. The behavior I'm seeing with mypy does not match its [documentation for overloaded calls](https://mypy.readthedocs.io/en/stable/more_types.html?highlight=%40overload#type-checking-calls-to-overloads).

**Comment by erictraut** [comment]
I've confirmed via the mypy source code that this is intended behavior. The mypy documentation is incomplete / incorrect in describing its behavior for overload matching. Pyright's current behavior matches the documented behavior of mypy, but I think it's reasonable to match its actual behavior for overload matching.

**Comment by erictraut** [comment]
This will be included in the next release.

**Comment by erictraut** [comment]
@Dr-Irv, with this change in place, I'm getting 12 new test failures in `pandas-stubs`. I'm investigating the cause. It appears to be due to some inconsistent and difficult-to-explain behaviors in mypy.

```python
from typing import Any
import numpy
import numpy.typing
import pandas as pd

td = pd.Timedelta()


def test_pandas_timedelta_1(
    nd: numpy.ndarray[Any, numpy.dtype[numpy.timedelta64]]
) -> None:
    reveal_type(nd)  # Mypy: numpy.ndarray[Any, numpy.dtype[numpy.timedelta64]]
    comp = td == nd
    reveal_type(comp)  # Mypy: Any


def test_pandas_timedelta_2() -> None:
    nd: numpy.ndarray[Any, numpy.dtype[numpy.timedelta64]]
    nd = numpy.array([1, 2, 3], dtype="timedelta64[D]")
    reveal_type(nd)  # Mypy: numpy.ndarray[Any, numpy.dtype[numpy.timedelta64]]
    comp = td == nd
    reveal_type(comp)  # Mypy: numpy.ndarray[Any, numpy.dtype[numpy.bool_]]
```

I think I'll revert this change until I better understand the underlying cause of the test breaks in pandas-stubs.

**Comment by Dr-Irv** [comment]
@erictraut I'm a little confused here.  Why would a change in `pyright` cause failures in using `mypy` to test `pandas-stubs` ?

Or did you try to change the `pandas-stubs` code to test the new `pyright` feature?

**Comment by Dr-Irv** [comment]
Not sure if this helps, but the following code (which uses the numpy suggested way of declaring types) passes mypy just fine:
```python
import numpy as np
import numpy.typing as npt
import pandas as pd


td = pd.Timedelta()


def test_pandas_timedelta_1(nd: npt.NDArray[np.timedelta64]) -> None:
    reveal_type(nd)  # Mypy: numpy.ndarray[Any, numpy.dtype[numpy.timedelta64]]
    comp = td == nd
    reveal_type(comp)  # Mypy: numpy.ndarray[Any, numpy.dtype[numpy.bool_]]


def test_pandas_timedelta_2() -> None:
    nd: npt.NDArray[np.timedelta64]
    nd = np.array([1, 2, 3], dtype="timedelta64[D]")
    reveal_type(nd)  # Mypy: numpy.ndarray[Any, numpy.dtype[numpy.timedelta64]]
    comp = td == nd
    reveal_type(comp)  # Mypy: numpy.ndarray[Any, numpy.dtype[numpy.bool_]]
```

**Comment by erictraut** [comment]
After spending several days trying to replicate mypy's overload behavior, I've come to the conclusion that there's really no way to replicate it exactly. Mypy's behavior is not precise, and its behaviors are dependent on implementation-specific internals that would be nearly impossible to replicate in other type checkers.

I took several runs at the problem, and each time pyright's behavior achieved parity with mypy's behavior in one area, it diverged in other areas. The end result would be a bunch of low-value churn for code bases that use pyright for type checking.

I incorporated a few small changes to pyright's overload evaluation behavior, but I don't plan to make any broader changes at this time.

If the typing community concludes that it's important for type checkers to have the same overload evaluation behavior, we should specify the algorithm in detail and provide a comprehensive test suite to validate the spec. I think I could fully document pyright's current algorithm, but it would be difficult because of the complexity. Mypy's current algorithm would be nearly impossible to specify in full, especially where `Any` types come into play, and I think it would be difficult to defend as a spec because it results in many demonstrably inconsistent behaviors.
